app/models/payment_order.rb

```diff
@@ -0,0 +1,102 @@
+class PaymentOrder < ApplicationRecord
+  include Versions
+  include ActionView::Helpers::NumberHelper
+
+  PAYMENT_INTERMEDIARIES = ENV['payments_intermediaries'].to_s.strip.split(', ').freeze
+  PAYMENT_BANKLINK_BANKS = ENV['payments_banks'].to_s.strip.split(', ').freeze
+  INTERNAL_PAYMENT_METHODS = %w[admin_payment system_payment].freeze
+  PAYMENT_METHODS = [PAYMENT_INTERMEDIARIES, PAYMENT_BANKLINK_BANKS,
+                     INTERNAL_PAYMENT_METHODS].flatten.freeze
+  CUSTOMER_PAYMENT_METHODS = [PAYMENT_INTERMEDIARIES, PAYMENT_BANKLINK_BANKS].flatten.freeze
+
+  belongs_to :invoice, optional: false
+
+  validate :invoice_cannot_be_already_paid, on: :create
+  validate :supported_payment_method
+
+  enum status: { issued: 'issued', paid: 'paid', cancelled: 'cancelled',
+                 failed: 'failed' }
+
+  attr_accessor :return_url, :response_url
+
+  def self.supported_methods
+    supported = []
+
+    PAYMENT_METHODS.each do |method|
+      class_name = ('PaymentOrders::' + method.camelize).constantize
+      raise(NoMethodError, class_name) unless class_name < PaymentOrder
+
+      supported << class_name
+    end
+
+    supported
+  end
+
+  def self.create_with_type(type:, invoice:)
```

Comment:
In Rails way methods like #new usually generates new objects, while #create - generates and saves.
This method is called #create_*, but behaves like #new.

Please, consider renaming or changing the behaviour, as you refactors this - there is no need to drag old fallacies in a bright new future =).